packages/flutter_tools/lib/src/commands/run.dart

```diff
@@ -610,6 +610,13 @@ class RunCommand extends RunCommandBase {
     webMode = featureFlags.isWebEnabled &&
       devices!.length == 1  &&
       await devices!.single.targetPlatform == TargetPlatform.web_javascript;
+
+    final String? flavor = stringArg('flavor');
+    final bool flavorsSupportedOnEveryDevice = devices!
+      .every((Device device) => device.supportsFlavors);
+    if (flavor != null && !flavorsSupportedOnEveryDevice) {
+      throwToolExit('--flavor is only supported for Android, macOS, and iOS devices.');
```

Comment:
A few thoughts:

1. The API here of `List<Device>? devices` here is unfortunate, but I won't ask you to refactor this in this PR. I'm guessing `late final Future<List<Devices>> devices = findAllTargetDevices();` would probably be more readable (although it's unfortunate that `findAllTargetDevices()` has hidden dependencies, but alas)
2. I'm guessing that here `devices` will be only devices that: are available AND we have determined we will be running the app on, right? In which case, it seems like this validation logic is correct.
3. Do we also need validation logic for `flutter test`

Reply:
> 2\. I'm guessing that here `devices` will be only devices that: are available AND we have determined we will be running the app on, right? In which case, it seems like this validation logic is correct.

This was my understanding from the last time I read the code. It also makes sense with the name of `findAllTargetDevices`.

> 3\. Do we also need validation logic for flutter test

This seems like a good idea. I first need to figure out how device selection works for `flutter test`.

Reply:
I've added the validation for `flutter test`. It is unclear to me whether or not this validation should itself be under test. Given its simplicity and that most of the neighboring `throwToolExits` also do not seem to be under test, I've elected to not create an integration test for this (perhaps I'm succumbing to broken window theory here). I did manually test this.